lgc/state/PipelineState.cpp

```diff
@@ -567,36 +576,43 @@ const ResourceNode *PipelineState::findPushConstantResourceNode() const {
+    if (nodeType == ResourceNodeType::Unknown || candidateType == nodeType)
+      return true;
+
+    if (nodeType == ResourceNodeType::DescriptorBuffer &&
+        (candidateType == ResourceNodeType::DescriptorBufferCompact || candidateType == ResourceNodeType::InlineBuffer))
+      return true;
+
+    if ((nodeType == ResourceNodeType::DescriptorResource || nodeType == ResourceNodeType::DescriptorTexelBuffer ||
+         nodeType == ResourceNodeType::DescriptorSampler) &&
+        (candidateType == ResourceNodeType::DescriptorCombinedTexture ||
+         candidateType == ResourceNodeType::DescriptorYCbCrSampler))
+      return true;
+
+    return false;
+  };
+
   for (const ResourceNode &node : getUserDataNodes()) {
     if (node.type == ResourceNodeType::DescriptorTableVaPtr) {
-      if (nodeType == node.type) {
-        if (!node.innerTable.empty() && node.innerTable[0].set == descSet)
+      // Check inner nodes.
+      bool isFirstChild = true;
+      for (const ResourceNode &innerNode : node.innerTable) {
+        if (isFirstChild && nodeType == ResourceNodeType::DescriptorTableVaPtr && innerNode.set == descSet)
           return {&node, &node};
-      } else {
-        for (const ResourceNode &innerNode : node.innerTable) {
-          if (innerNode.set == descSet && innerNode.binding == binding) {
-            if (nodeType == ResourceNodeType::Unknown || nodeType == innerNode.type ||
-                (nodeType == ResourceNodeType::DescriptorBuffer &&
-                 (innerNode.type == ResourceNodeType::DescriptorBufferCompact ||
-                  innerNode.type == ResourceNodeType::InlineBuffer)) ||
-                ((innerNode.type == ResourceNodeType::DescriptorCombinedTexture ||
-                  innerNode.type == ResourceNodeType::DescriptorYCbCrSampler) &&
-                 (nodeType == ResourceNodeType::DescriptorResource ||
-                  nodeType == ResourceNodeType::DescriptorTexelBuffer ||
-                  nodeType == ResourceNodeType::DescriptorSampler)))
-              return {&node, &innerNode};
-          }
-        }
+        isFirstChild = false;
+
+        if (innerNode.set == descSet && innerNode.binding == binding && IsNodeTypeCompatible(innerNode.type))
+          return {&node, &innerNode};
       }
-    } else if (node.set == descSet && node.binding == binding) {
-      if (nodeType == ResourceNodeType::Unknown || nodeType == node.type ||
-          (nodeType == ResourceNodeType::DescriptorBuffer && node.type == ResourceNodeType::DescriptorBufferCompact) ||
-          ((node.type == ResourceNodeType::DescriptorCombinedTexture ||
-            node.type == ResourceNodeType::DescriptorYCbCrSampler) &&
-           (nodeType == ResourceNodeType::DescriptorResource || nodeType == ResourceNodeType::DescriptorTexelBuffer ||
-            nodeType == ResourceNodeType::DescriptorSampler)))
-        return {&node, &node};
+      return {nullptr, nullptr};
```

Comment:
This looks like you return "not found" if the first `ResourceNodeType::DescriptorTableVaPtr` is not the one you are looking for.  However, the node you want could come in a later top level node.  Are you sure this is correct?